website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+    print("---TRANSLATING INPUT---")
+    messages = state["messages"]
+    user_input = messages[-1][1]  # Get the latest user input
+
+    # Translation prompt for the model
+    translation_prompt = f"""
+Translate the following text to English. If the text is already in English, repeat it exactly without any additional explanation.
+
+Text:
+{user_input}
+"""
+    # Call the OpenAI LLM to translate the text
+    translated_response = llm.invoke(translation_prompt)
+    translated_text = translated_response.content.strip()  # Access the 'content' attribute and strip any extra spaces
+    state["translated_input"] = translated_text  # Save the translated input
+    print(f"Translated Input: {translated_text}")
+
+    return state
+```
+
+#### **2. Safety Check**
+
+The `safety_check` node ensures user inputs are free from harmful, malicious, or potentially destructive SQL commands and inappropriate content.
+
+**Purpose:** Filter out unsafe or harmful queries.
+
+**Process:**
+
+- Scans for disallowed SQL operations (*i.e.*, `CREATE`, `DELETE`, `DROP`, `INSERT`, `UPDATE`, `ALTER`, `TRUNCATE`, `EXEC`, `EXECUTE`).
+
+- Uses the LLM to analyze the input for toxic or inappropriate content.
+
+- Updates the state with an error flag and feedback if the query is unsafe.
+
+**Key Considerations:**
+
+- Rejects queries containing disallowed SQL keywords.
+
+- Provides transparent feedback for flagged inputs.
+
+- Ensures the query is free of offensive or harmful language.
+
+**Examples:**
+
+1. **Disallowed Operations:**
+
+- **Input:** *"DROP TABLE customers;"*
+
+- **Response:** *"Your query contains disallowed SQL operations and cannot be processed."*
+
+- **Input:** *"SELECT * FROM orders;"*
+
+- **Response:** *"Query allowed."*
+
+2. **Inappropriate Content:**
+
+- **Input:** *"Show me orders where customers have names like 'John the Idiot';"*
+
+- **Response:** *"Your query contains inappropriate content and cannot be processed."*
+
+- **Input:** *"Find total sales by region."*
+
+- **Response:** *"Input is safe to process."*
+
+**Code:**
+
+```python
+def safety_check(state: GraphState):
+    print("---PERFORMING SAFETY CHECK---")
+    translated_input = state["translated_input"]
+    messages = state["messages"]
+    error = "no"
+
+    # List of disallowed SQL operations (e.g., DELETE, DROP)
+    disallowed_operations = ['CREATE', 'DELETE', 'DROP', 'INSERT', 'UPDATE', 'ALTER', 'TRUNCATE', 'EXEC', 'EXECUTE']
+    pattern = re.compile(r'\b(' + '|'.join(disallowed_operations) + r')\b', re.IGNORECASE)
+
+    # Check if the input contains disallowed SQL operations
+    if pattern.search(translated_input):
+        print("Input contains disallowed SQL operations. Halting the workflow.")
+        error = "yes"
+        messages += [("assistant", "Your query contains disallowed SQL operations and cannot be processed.")]
+    else:
+        # Check if the input contains inappropriate content
+        safety_prompt = f"""
```

Comment:
curious about using pattern search on the natural language input—wouldn't checking for disallowed SQL operations make more sense after the SQL is generated?

i.e. what would happen if the natural language input is "please get rid of the customers table." It doesn't look like checks for disallowed operations are run again after the sql is generated, so might it be possible that the system would generate and run a drop table command as long as the user didn't explicitly say "drop table"?